docs/tutorial/YAML-configuration.md
In UMIche, all calculation methods rely on a `YAML` file for parameter initialisation.

It defines 5 sections

- [x] `work_dir` - working directory
- [x] `trimmed` - trimming FastQ reads and extracting barcodes and UMIs
- [x] `fixed` - single-valued simulation parameters
- [x] `varied` - varying-valued simulation parameters
- [x] `dedup` - UMI deduplication parameters

# Monomer UMI pipeline settings

``` shell
work_dir: /mnt/d/Document/Programming/Python/umiche/umiche/data/simu/mclumi/
#work_dir: D:/Document/Programming/Python/umiche/umiche/data/simu/general/seq_errs/
# work_dir data/simu/tree/trimer/
# work_dir data/simu/monomer/pcr8/
# work_dir data/simu/trimer/pcr8/
# work_dir data/simu/dimer/pcr8/
# work_dir data/simu/dimer/treepcr22_250/
# work_dir data/simu/dimer/pcr8_mono24/

trimmed:
  fastq:
    fpn: None
    trimmed_fpn: None

  umi_1:
    len: 10

  seq:
    len: 100

  read_struct: 'umi_1'


fixed:
  pcr_num: 8
  pcr_err: 0.00001
  seq_err: 0.001
  ampl_rate: 0.85
  seq_dep: 400
  umi_num: 50
  permutation_num: 2
  umi_unit_pattern: 1
  umi_unit_len: 10
  seq_sub_spl_rate: 0.333
  sim_thres: 3


varied:
  pcr_nums: [ # pcr_nums_err_2d_spl0.33
    1, 2, 3,
    4, 5, 6, 7,
    8, 9, 10, 11, 12,
    13, 14, 15, 16,
#    17, 18
#    17, 18, 19, 20,
  ]
  pcr_errs: [
    0.00001,
    0.000025,
    0.00005,
    0.000075,
    0.0001,
    0.00025,
    0.0005,
    0.00075,
    0.001,
    0.0025,
    0.005,
    0.0075,
    0.01,
#    0.025,
    0.05,
#    0.075,
#    0.1,
#    0.2,
#    0.3,
  ]
  seq_errs: [
    0.00001,
    0.000025,
    0.00005,
    0.000075,
    0.0001,
    0.00025,
    0.0005,
    0.00075,
    0.001,
    0.0025,
    0.005,
    0.0075,
    0.01,
    0.025,
    0.05,
    0.075,
    0.1,
#    0.2,
#    0.3,
  ]
  ampl_rates: [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
  umi_lens: [6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
#  umi_lens: [6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36]
#  umi_nums: [50, 250, 450, 650, 850, 1050]
  umi_nums: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45]
  seq_deps: [100, 200, 500, 600, 800, 1000, 2000, 3000, 5000 ]
#  seq_deps: [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]


dedup:
  dbscan_eps: 1.5 # 1.5
  dbscan_min_spl: 1
  birch_thres: 1.8 # 1.8
  birch_n_clusters: None
  hdbscan_min_spl: 3
  aprop_preference: None
  aprop_random_state: 0

  ed_thres: 1
  mcl_fold_thres: 1.6 # 1.6
  iter_num: 100

#  inflat_val: 2.7 # 1.1 2.7
#  exp_val: 2 # 2 3
#
  inflat_val: [1.1, 2.7, 3.6]
  exp_val: 2

#  exp_val: [2, 3, 4]

  # mcl_ed trace!!!
#  ed_thres: 1
#  mcl_fold_thres: 2 # 1.6
#  inflat_val: 2.7 # 1.1 2.7
#  exp_val: 2 # 2 3
#  iter_num: 100

  # pcr_nums
  # mcl_inflat: 2.3
  # mcl_exp: 2
  # mcl_fold_thres: 1
```


# Homotrimer UMI pipeline settings

``` shell
#work_dir: d:/Document/Programming/Python/umiche/umiche/data/simu/umiche/trimer/
work_dir: /mnt/d/Document/Programming/Python/umiche/umiche/data/simu/umiche/trimer/

trimmed:
  fastq:
    fpn: None
    trimmed_fpn: None

  umi_1:
    len: 36

  seq:
    len: 100

  read_struct: 'umi_1'


fixed:
  pcr_num: 8
  pcr_err: 0.00001
  seq_err: 0.001
  ampl_rate: 0.85
  seq_dep: 400
  umi_num: 50
  permutation_num: 10
  umi_unit_pattern: 3
  umi_unit_len: 12
  seq_sub_spl_rate: 0.333
  sim_thres: 3


varied:
  pcr_nums: [ # pcr_nums_err_2d_spl0.33
    1, 2, 3,
    4, 5, 6, 7,
    8, 9, 10, 11, 12,
    13, 14, 15, 16,
#    17, 18
#    17, 18, 19, 20,
  ]
  pcr_errs: [
    0.00001,
    0.000025,
    0.00005,
    0.000075,
    0.0001,
    0.00025,
    0.0005,
    0.00075,
    0.001,
    0.0025,
    0.005,
    0.0075,
    0.01,
#    0.025,
    0.05,
#    0.075,
#    0.1,
#    0.2,
#    0.3,
  ]
  seq_errs: [
    0.00001,
    0.000025,
    0.00005,
    0.000075,
    0.0001,
    0.00025,
    0.0005,
    0.00075,
    0.001,
    0.0025,
    0.005,
    0.0075,
    0.01,
    0.025,
    0.05,
    0.075,
    0.1,
    0.2,
#    0.3,
  ]
  ampl_rates: [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
  umi_lens: [6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
#  umi_lens: [6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36]
#  umi_nums: [50, 250, 450, 650, 850, 1050]
  umi_nums: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45]
  seq_deps: [100, 200, 500, 600, 800, 1000, 2000, 3000, 5000 ]
#  seq_deps: [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]


dedup:
  dbscan_eps: 1.5 # 1.5
  dbscan_min_spl: 1
  birch_thres: 1.8 # 1.8
  birch_n_clusters: None
  hdbscan_min_spl: 3
  aprop_preference: None
  aprop_random_state: 0

  ed_thres: 1
  mcl_fold_thres: 1.6 # 1.6
  inflat_val: 2.7 # 1.1 2.7
  exp_val: 2 # 2 3
  iter_num: 100

  # mcl_ed trace!!!
#  ed_thres: 1
#  mcl_fold_thres: 2 # 1.6
#  inflat_val: 2.7 # 1.1 2.7
#  exp_val: 2 # 2 3
#  iter_num: 100

  # pcr_nums
  # mcl_inflat: 2.3
  # mcl_exp: 2
  # mcl_fold_thres: 1
```

umiche/deduplicate/pipeline/Config.py
__version__ = "v1.0"
__copyright__ = "Copyright 2024"
__license__ = "MIT"
__developer__ = "Jianfeng Sun"
__maintainer__ = "Jianfeng Sun"
__email__="jianfeng.sunmt@gmail.com"
__lab__ = "Cribbslab"

import sys
import numpy as np
# from simreadflow.simulate.dispatcher.single.General import general as simugeneral
# from Path import to


class config:

    def __init__(self, ):
        # ### /*** block. general ***/
        # self.permutation_num = 10
        #
        # self.umi_unit_len_fixed = 10
        # self.umi_num_fixed = 50
        # self.pcr_num_fixed = 10
        # self.pcr_err_fixed = 1e-3
        # self.seq_err_fixed = 1e-3
        # self.ampl_rate_fixed = 0.85
        # self.sim_thres_fixed = 3
        # self.seq_sub_spl_rate = 1
        #
        # self.ampl_rates = np.linspace(0.1, 1, 10)
        # self.umi_unit_lens = np.arange(6, 36 + 1, 1)
        # self.umi_nums = np.arange(20, 140 + 20, 20)
        # self.pcr_nums = np.arange(1, 20 + 1, 1)
        # self.pcr_errs, self.seq_errs = self.errors()
        # print(self.pcr_errs, self.seq_errs)

        # ### /*** block. bulk ***/
        self.permutation_num = 10
        self.umi_unit_pattern = 1
        self.umi_unit_len_fixed = 36
        self.umi_num_fixed = 50
        self.pcr_num_fixed = 8
        self.pcr_err_fixed = 1e-3
        self.seq_err_fixed = 1e-3
        self.ampl_rate_fixed = 0.85
        self.sim_thres_fixed = 3
        self.seq_sub_spl_rate = 1
        self.working_dir = ''

        self.ampl_rates = np.linspace(0.1, 1, 10)
        self.umi_unit_lens = np.arange(6, 36 + 1, 1)
        self.umi_nums = np.arange(20, 140 + 20, 20)
        self.pcr_nums = np.arange(1, 20 + 1, 1)
        self.pcr_errs, self.seq_errs = self.errors()
        # self.pcr_errs = [1e-05, 2.5e-05, 5e-05, 7.5e-05, 0.0001, 0.00025, 0.0005, 0.00075, 0.001, 0.0025, 0.005, 0.0075, 0.01]
        print(self.pcr_errs)
        print(self.seq_errs)

        self.metrics = {
            'pcr_nums': self.pcr_nums,
            'pcr_errs': self.pcr_errs,
            'seq_errs': self.seq_errs,
            'ampl_rates': self.ampl_rates,
            'umi_lens': self.umi_unit_lens,
        }
        self.fastq_fn_pref = {
            'pcr_nums': 'pcr_',
            'pcr_errs': 'pcr_err_',
            'seq_errs': 'seq_err_',
            'ampl_rates': 'ampl_rate_',
            'umi_lens': 'umi_len_',
        }

    def errors(self, ):
        pcr_errs = []
        seq_errs = []
        e = 1e-5
        while e < 3e-1:
            pcr_errs.append(e)
            seq_errs.append(e)
            if 5 * e < 3e-1:
                pcr_errs.append(2.5 * e)
                pcr_errs.append(5 * e)
                pcr_errs.append(7.5 * e)
                seq_errs.append(2.5 * e)
                seq_errs.append(5 * e)
                seq_errs.append(7.5 * e)
            e = 10 * e
        pcr_errs.append(0.125)
        seq_errs.append(0.125)
        pcr_errs.append(0.15)
        seq_errs.append(0.15)
        pcr_errs.append(0.2)
        seq_errs.append(0.2)
        pcr_errs.append(0.225)
        seq_errs.append(0.225)
        pcr_errs.append(0.25)
        seq_errs.append(0.25)
        pcr_errs.append(0.3)
        seq_errs.append(0.3)
        # print(pcr_errs)
        # print(seq_errs)
        return pcr_errs, seq_errs

    def pcrNums(self, ):
        for pn in range(self.permutation_num):
            simu_params = {
                'init_seq_setting': {
                    'seq_num': self.umi_num_fixed,
                    'umi_unit_pattern': 1,
                    'umi_unit_len': self.umi_unit_len_fixed,
                    # 'seq_len': self.seq_len_fixed - self.umi_unit_len_fixed,
                    'is_seed': True,
                    'working_dir': to('data/simu/monomer/general/1/pcr_num/permute_') + str(pn) + '/',
                    'is_sv_umi_lib': True,
                    'umi_lib_fpn': to('data/simu/monomer/general/1/pcr_num/permute_') + str(pn) + '/umi.txt',
                    # 'is_sv_seq_lib': True,
                    # 'seq_lib_fpn': to('data/simu/monomer/general/1/pcr_num/permute_') + str(pn) + '/seq.txt',
                    'condis': ['umi'],
                    'sim_thres': self.sim_thres_fixed,
                    'permutation': pn,
                },
                'ampl_rate': self.ampl_rate_fixed,
                'pcr_nums': self.pcr_nums,
                'err_num_met': 'nbinomial',
                'pcr_error': self.pcr_err_fixed,
                'seq_error': self.seq_err_fixed,
                'seq_sub_spl_rate': self.seq_sub_spl_rate,
                'use_seed': False,
                'seed': None,
                'write': {
                    'fastq_fp': to('data/simu/monomer/general/1/pcr_num/permute_') + str(pn) + '/',
                    'fastq_fn': '',
                }
            }
            p = simugeneral(simu_params)
            print(p.ondemandPCRNums())
        return

    def pcrErrs(self, ):
        for pn in range(self.permutation_num):
            simu_params = {
                'init_seq_setting': {
                    'seq_num': self.umi_num_fixed,
                    'umi_unit_pattern': 3,
                    'umi_unit_len': self.umi_unit_len_fixed,
                    # 'seq_len': self.seq_len_fixed - self.umi_unit_len_fixed,
                    'is_seed': True,
                    # ### /*** block. general ***/
                    # 'working_dir': to('data/simu/monomer/general/1/pcr_err/permute_') + str(pn) + '/',
                    # 'is_sv_umi_lib': True,
                    # 'umi_lib_fpn': to('data/simu/monomer/general/1/pcr_err/permute_') + str(pn) + '/umi.txt',

                    # # ### /*** block. dimer ***/
                    # 'working_dir': to('data/simu/dimer/pcr_err/permute_') + str(pn) + '/',
                    # 'is_sv_umi_lib': True,
                    # 'umi_lib_fpn': to('data/simu/dimer/pcr_err/permute_') + str(pn) + '/umi.txt',

                    # ### /*** block. trimer ***/
                    'working_dir': to('data/simu/trimer/pcr_err/permute_') + str(pn) + '/',
                    'is_sv_umi_lib': True,
                    'umi_lib_fpn': to('data/simu/trimer/pcr_err/permute_') + str(pn) + '/umi.txt',

                    'condis': ['umi'],
                    'sim_thres': self.sim_thres_fixed,
                    'permutation': pn,
                },
                'ampl_rate': self.ampl_rate_fixed,
                'pcr_num': self.pcr_num_fixed,
                'err_num_met': 'nbinomial',
                'pcr_errors': self.pcr_errs,
                'seq_error': self.seq_err_fixed,
                'seq_sub_spl_rate': self.seq_sub_spl_rate,
                'use_seed': False,
                'seed': None,
                'write': {
                    'fastq_fp': to('data/simu/trimer/pcr_err/permute_') + str(pn) + '/',
                    'fastq_fn': '',
                }
            }
            p = simugeneral(simu_params)
            print(p.ondemandPCRErrs())
        return

    def seqErrs(self, ):
        sys.stdout = open(self.working_dir + 'log.txt', 'w')
        for pn in range(self.permutation_num):
            simu_params = {
                'init_seq_setting': {
                    'seq_num': self.umi_num_fixed,
                    'umi_unit_pattern': self.umi_unit_pattern,
                    'umi_unit_len': self.umi_unit_len_fixed,
                    # 'seq_len': self.seq_len_fixed - self.umi_unit_len_fixed,
                    'is_seed': True,

                    'working_dir': self.working_dir + 'seq_errs/permute_' + str(pn) + '/',
                    'is_sv_umi_lib': True,
                    'umi_lib_fpn': self.working_dir + 'seq_errs/permute_' + str(pn) + '/umi.txt',

                    'condis': ['umi'],
                    'sim_thres': self.sim_thres_fixed,
                    'permutation': pn,
                },
                'ampl_rate': self.ampl_rate_fixed,
                'pcr_num': self.pcr_num_fixed,
                'err_num_met': 'nbinomial',
                'pcr_error': self.pcr_err_fixed,
                'seq_errors': self.seq_errs,
                'seq_sub_spl_rate': self.seq_sub_spl_rate,
                'use_seed': False,
                'seed': None,
                'write': {
                    'fastq_fp': self.working_dir + 'seq_errs/permute_' + str(pn) + '/',
                    'fastq_fn': '',
                }
            }
            p = simugeneral(simu_params)
            print(p.ondemandSeqErrs())
        sys.stdout.close()
        return

    def umiLens(self, ):
        for pn in range(self.permutation_num):
            simu_params = {
                'init_seq_setting': {
                    'seq_num': self.umi_num_fixed,
                    'umi_unit_pattern': 1,
                    'umi_unit_lens': self.umi_unit_lens,
                    # 'seq_len': self.seq_len_fixed,
                    'is_seed': True,
                    'working_dir': to('data/simu/monomer/general/1/umi_len/permute_') + str(pn) + '/',
                    'is_sv_umi_lib': True,
                    'umi_lib_fp': to('data/simu/monomer/general/1/umi_len/permute_') + str(pn) + '/',
                    # 'is_sv_seq_lib': True,
                    # 'seq_lib_fpn': to('data/simu/monomer/general/1/umi_len/permute_') + str(pn) + '/',
                    'condis': ['umi'],
                    'sim_thres': self.sim_thres_fixed,
                    'permutation': pn,
                },
                'ampl_rate': self.ampl_rate_fixed,
                'pcr_num': self.pcr_num_fixed,
                'err_num_met': 'nbinomial',
                'pcr_error': self.pcr_err_fixed,
                'seq_error': self.seq_err_fixed,
                'seq_sub_spl_rate': self.seq_sub_spl_rate,
                'use_seed': False,
                'seed': None,
                'write': {
                    'fastq_fp': to('data/simu/monomer/general/1/umi_len/permute_') + str(pn) + '/',
                    'fastq_fn': '',
                }
            }
            p = simugeneral(simu_params)
            print(p.ondemandUMILens())
        return

    def amplRates(self, ):
        for pn in range(self.permutation_num):
            simu_params = {
                'init_seq_setting': {
                    'seq_num': self.umi_num_fixed,
                    'umi_unit_pattern': 1,
                    'umi_unit_len': self.umi_unit_len_fixed,
                    # 'seq_len': self.seq_len_fixed - self.umi_unit_len_fixed,
                    'is_seed': True,
                    'working_dir': to('data/simu/monomer/general/1/ampl_rate/permute_') + str(pn) + '/',
                    'is_sv_umi_lib': True,
                    'umi_lib_fpn': to('data/simu/monomer/general/1/ampl_rate/permute_') + str(pn) + '/umi.txt',
                    # 'is_sv_seq_lib': True,
                    # 'seq_lib_fpn': to('data/simu/monomer/general/1/ampl_rate/permute_') + str(pn) + '/seq.txt',
                    'condis': ['umi'],
                    'sim_thres': self.sim_thres_fixed,
                    'permutation': pn,
                },
                'ampl_rates': self.ampl_rates,
                'pcr_num': self.pcr_num_fixed,
                'err_num_met': 'nbinomial',
                'pcr_error': self.pcr_err_fixed,
                'seq_error': self.seq_err_fixed,
                'seq_sub_spl_rate': self.seq_sub_spl_rate,
                'use_seed': False,
                'seed': None,
                'write': {
                    'fastq_fp': to('data/simu/monomer/general/1/ampl_rate/permute_') + str(pn) + '/',
                    'fastq_fn': '',
                }
            }
            p = simugeneral(simu_params)
            print(p.ondemandAmplRates())
        return


if __name__ == "__main__":
    p = general(
        working_dir=to('data/simu/trimer/pcr8/')
        # working_dir=to('data/simu/trimer/pcr8_mono24/')
    )

    print(p.seqErrs())

    # print(p.pcrNums())

    # print(p.pcrErrs())

    # print(p.umiLens())

    # print(p.amplRates())